frontend/extensions/helpers.lua

```diff
@@ -0,0 +1,275 @@
+
+	else
+		UIManager:show(InfoMessage:new { text = message, icon_file = 'resources/info-bug.png', timeout = timeout, dismiss_callback = dismiss_callback })
+	end
+end
+
+function Helpers:alertDebugOnce(message, timeout, dismiss_callback)
+
+	if not self.debug_once_used then
+
+		self:alertDebug(message, timeout, dismiss_callback)
+
+		self.debug_once_used = true
+	end
+end
+
+-- use timeout = nil for second argument when calling for no timeout:
+function Helpers:alertError(message, timeout, dismiss_callback)
+
+	if dismiss_callback == nil then
+		UIManager:show(InfoMessage:new { text = message, icon_file = 'resources/info-error.png', timeout = timeout })
+
+	else
+		UIManager:show(InfoMessage:new { text = message, icon_file = 'resources/info-error.png', timeout = timeout, dismiss_callback = dismiss_callback })
+	end
+end
+
+-- use timeout = nil for second argument when calling for no timeout:
+function Helpers:alertInfo(message, timeout, dismiss_callback)
+
+	if dismiss_callback == nil then
+		UIManager:show(InfoMessage:new { text = message, timeout = timeout })
+
+	else
+		UIManager:show(InfoMessage:new { text = message, timeout = timeout, dismiss_callback = dismiss_callback })
+	end
+end
+
+-- use timeout = nil for second argument when calling for no timeout:
+function Helpers:alertNoIcon(message, timeout)
+
+	UIManager:show(InfoMessage:new { text = message, show_icon = false, timeout = timeout })
+end
+
+function Helpers:shuffle(tbl)
+	for i = #tbl, 2, -1 do
+		local j = math.random(i)
+		tbl[i], tbl[j] = tbl[j], tbl[i]
+	end
+end
+
+function Helpers:split(str, pat)
+	local t = {}  -- NOTE: use {n = 0} in Lua-5.0
+	local fpat = "(.-)" .. pat
+	local last_end = 1
+	local s, e, cap = str:find(fpat, 1)
+	while s do
+		if s ~= 1 or cap ~= "" then
+			table.insert(t, cap)
+		end
+		last_end = e + 1
+		s, e, cap = str:find(fpat, last_end)
+	end
+	if last_end <= #str then
+		cap = str:sub(last_end)
+		table.insert(t, cap)
+	end
+	return t
+end
+
+function Helpers:split_path(str)
+	return self:split(str, '[\\/]+')
+end
+
+function Helpers:table_sort_by_text_prop(subject)
+
+	table.sort(subject, function(v1, v2)
+		return v1.text < v2.text
+	end)
+end
+
+function Helpers:table_sort_alphabetically(subject)
+
+	table.sort(subject, function(v1, v2)
+		return v1 < v2
+	end)
+end
+
+function Helpers:ucfirst(str)
+	return (str:gsub("^%l", string.upper))
+end
+
+function Helpers:getMiddleDialogWidth()
+
+	local orientation = Screen:getScreenMode()
+	local iwidth = (Screen:getWidth() / 2) + 20
+
+	if orientation == 'portrait' then
+		iwidth = Screen:getWidth() - 80
+	end
+
+	return iwidth
+end
+
+function Helpers:getWideDialogWidth()
+
+	local orientation = Screen:getScreenMode()
+	local iwidth = Screen:getWidth() - 120
+
+	if orientation == 'portrait' then
+		iwidth = Screen:getWidth() - 80
+	end
+
+	return iwidth
+end
+
+function Helpers:round(x)
+
+	if x % 2 ~= 0.5 then
```

Comment:
As an aside, this is not how I round. 1.5 rounds to 2, not to 1.

Reply:
When you run the code, you'll see that it rounds 1.5 nicely to 2:
 1.5 % 2 = 0 => so not 0.5, so return math.floor(1.5 + 0.5) => 2

Reply:
1.5 mod 2 = 1.5, so you return 1.5 - 0.5 = 1

0.7 mod 2 = 0.7, so you return 0.7 - 0.5 = 0.2

Reply:
Wait, you and/or I confused me, but this is how your function behaves pasted into a Lua console. >_>

```
> print (Helpers:round(1.5))
1
> print (Helpers:round(2))
1.5
> print (Helpers:round(2.5))
2
> print (Helpers:round(0.7))
0.2
> print (Helpers:round(2.5))
2
> print (Helpers:round(1.5))
1
```

In any case the exception makes no sense.

```
function round(x)

	if x % 2 ~= 0.5 then
		return math.floor(x + 0.5)
	end

	return x - 0.5
end

print(round(.5))
print(round(1.5))
```
```
$ lua bla.lua 
0
2
```

Reply:
Don't we already have a round in optmath?

Reply:
(Which should essentially be doing floor + 0.5).

Reply:
We do. GH may have helpfully hidden my comment saying exactly that right above. ;-)

Reply:
Why, yes we do, and yes it does ;).

https://github.com/koreader/koreader/blob/23862fd0ebd51f8a21ed4dd82861806f356f6128/frontend/optmath.lua#L43

Reply:
This one: https://github.com/koreader/koreader/pull/6255#discussion_r438648691

Reply:
Oh, actually, it might be wrong. Does floor(n + 0.5) == trunc(floor(n) + 0.5)?

Reply:
I don't think trunc is a thing in Lua, but that's probably the same thing. Why is that wrong?

Reply:
I dunno, but in C I'd have done floor(n) + 0.5f cast to an integer (so, truncated).

Reply:
My bad, I was assuming trunc(n + 0.5) as trunc(floor(n) + 0.5) looks completely meaningless to me.

Reply:
As in:

floor(1.9) = 1
trunc(1.9) = 1
trunc(1 + 0.5) = 1

Reply:
It'll be different for negative integers of course.

Reply:
Yeah, no, I still suck at maths ;D.

floor(n) + 0.5f essentially does a ceil, not a round.

TL;DR: Everything's fine :D.

Reply:
Never mind me, I still suck at maths ;D.

(i.e., floor(n + 0.5) works just fine ;)).

Reply:
And I was actually thinking of `(int) (n + 0.5f)`, and not the nonsense I mentioned earlier ^^.

Reply:
(So, basically, yeah, floor() :D).

Reply:
Difference illustrated in JS (roundToZero would be a much clearer name, cf. our [roundAwayFromZero](http://koreader.rocks/doc/modules/optmath.html#Math.roundAwayFromZero)):
```
Math.floor(-1.9)
-2
Math.trunc(-1.9)
-1
```